# Análisis de Canciones Favoritas en Spotify 

In [ ]:
Realizado por:
Adrian Marcelo Ballesteros Herrera
Miguel Aaron Castillo Ochoa

## Objetivos Generales

## Objetivos Específicos

## Marco Teórico

## Análisis de Dataset

## Modelos Propuestos

En este proyecto presentares tres distintos de modelos, en los cuales se compararan entre ellos mediante su forma sin penalización, con penalización de Ridge, Lasso y ElsaticNet.

### Modelo de Orden 3

In [1]:
import pandas as pd
import statsmodels.api as sm
from sklearn.linear_model import LinearRegression, Ridge, Lasso
import numpy as np
import matplotlib.pyplot as plt
from sklearn.metrics import r2_score
from sklearn.metrics import mean_squared_error
from sklearn.preprocessing import StandardScaler 
from sklearn.model_selection import train_test_split
from sklearn.linear_model import ElasticNet
from sklearn.preprocessing import PolynomialFeatures

In [4]:
# Leemos el dataset
df=pd.read_csv("Popular_Spotify_Songs (1).csv", encoding='latin1')
#agregamos enconding='latin 1', para que se pudiera leer el archivo pues no podia decodificar mediante el tipo utf-8
df.head()

track_name    artist(s)_name  artist_count  \
0  Seven (feat. Latto) (Explicit Ver.)  Latto, Jung Kook             2   
1                                 LALA       Myke Towers             1   
2                              vampire    Olivia Rodrigo             1   
3                         Cruel Summer      Taylor Swift             1   
4                       WHERE SHE GOES         Bad Bunny             1   

   released_year  released_month  released_day  in_spotify_playlists  \
0           2023               7            14                   553   
1           2023               3            23                  1474   
2           2023               6            30                  1397   
3           2019               8            23                  7858   
4           2023               5            18                  3133   

   in_spotify_charts    streams  in_apple_playlists  ...  bpm key   mode  \
0                147  141381703                  43  ...  125   B  Major   
1                 48  133716286                  48  ...   92  C#  Major   
2                113  140003974                  94  ...  138   F  Major   
3                100  800840817                 116  ...  170   A  Major   
4                 50  303236322                  84  ...  144   A  Minor   

  danceability_%  valence_% energy_% acousticness_%  instrumentalness_%  \
0             80         89       83             31                   0   
1             71         61       74              7                   0   
2             51         32       53             17                   0   
3             55         58       72             11                   0   
4             65         23       80             14                  63   

   liveness_%  speechiness_%  
0           8              4  
1          10              4  
2          31              6  
3          11             15  
4          11              6  

[5 rows x 24 columns]

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 953 entries, 0 to 952
Data columns (total 24 columns):
 #   Column                Non-Null Count  Dtype 
---  ------                --------------  ----- 
 0   track_name            953 non-null    object
 1   artist(s)_name        953 non-null    object
 2   artist_count          953 non-null    int64 
 3   released_year         953 non-null    int64 
 4   released_month        953 non-null    int64 
 5   released_day          953 non-null    int64 
 6   in_spotify_playlists  953 non-null    int64 
 7   in_spotify_charts     953 non-null    int64 
 8   streams               953 non-null    object
 9   in_apple_playlists    953 non-null    int64 
 10  in_apple_charts       953 non-null    int64 
 11  in_deezer_playlists   953 non-null    object
 12  in_deezer_charts      953 non-null    int64 
 13  in_shazam_charts      903 non-null    object
 14  bpm                   953 non-null    int64 
 15  key                   858 non-null    ob

In [6]:
df1=df.dropna()
df1.info()

<class 'pandas.core.frame.DataFrame'>
Index: 817 entries, 0 to 952
Data columns (total 24 columns):
 #   Column                Non-Null Count  Dtype 
---  ------                --------------  ----- 
 0   track_name            817 non-null    object
 1   artist(s)_name        817 non-null    object
 2   artist_count          817 non-null    int64 
 3   released_year         817 non-null    int64 
 4   released_month        817 non-null    int64 
 5   released_day          817 non-null    int64 
 6   in_spotify_playlists  817 non-null    int64 
 7   in_spotify_charts     817 non-null    int64 
 8   streams               817 non-null    object
 9   in_apple_playlists    817 non-null    int64 
 10  in_apple_charts       817 non-null    int64 
 11  in_deezer_playlists   817 non-null    object
 12  in_deezer_charts      817 non-null    int64 
 13  in_shazam_charts      817 non-null    object
 14  bpm                   817 non-null    int64 
 15  key                   817 non-null    object


In [7]:
df1=df1.drop(columns=["track_name","artist(s)_name","artist_count","released_month","released_day","in_apple_playlists","in_apple_charts","in_deezer_playlists","in_deezer_charts","in_shazam_charts","key","mode"])
df1

released_year  in_spotify_playlists  in_spotify_charts    streams  bpm  \
0             2023                   553                147  141381703  125   
1             2023                  1474                 48  133716286   92   
2             2023                  1397                113  140003974  138   
3             2019                  7858                100  800840817  170   
4             2023                  3133                 50  303236322  144   
..             ...                   ...                ...        ...  ...   
948           2022                   953                  0   91473363  144   
949           2022                  1180                  0  121871870  166   
950           2022                   573                  0   73513683   92   
951           2022                  1320                  0  133895612   97   
952           2022                   782                  2   96007391   90   

     danceability_%  valence_%  energy_%  acousticness_%  instrumentalness_%  \
0                80         89        83              31                   0   
1                71         61        74               7                   0   
2                51         32        53              17                   0   
3                55         58        72              11                   0   
4                65         23        80              14                  63   
..              ...        ...       ...             ...                 ...   
948              60         24        39              57                   0   
949              42          7        24              83                   1   
950              80         81        67               4                   0   
951              82         67        77               8                   0   
952              61         32        67              15                   0   

     liveness_%  speechiness_%  
0             8              4  
1            10              4  
2            31              6  
3            11             15  
4            11              6  
..          ...            ...  
948           8              3  
949          12              6  
950           8              6  
951          12              5  
952          11              5  

[817 rows x 12 columns]

Decidimios eliminar las columnas de "track_name"; "artist(s)_name"; "artist_count"; "released_month"; "released_day"; "key"; "mode", porque son variables que no nos ofrecen información relevante para nuestro analisis de que tipo de canciones son las favoritas para las personas hoy en día. También, como nos enfocamos en Spotify, eliminamos las columnas que nos hablan sobre otras aplicaciones de música. Por otra parte, no necesitamos la transfomación de datos dummies, pues no tenemos valores cualitativos.

In [8]:
df1["streams"] = pd.to_numeric(df1["streams"], errors='coerce')
df1=df1.dropna()
df1.info()

<class 'pandas.core.frame.DataFrame'>
Index: 816 entries, 0 to 952
Data columns (total 12 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   released_year         816 non-null    int64  
 1   in_spotify_playlists  816 non-null    int64  
 2   in_spotify_charts     816 non-null    int64  
 3   streams               816 non-null    float64
 4   bpm                   816 non-null    int64  
 5   danceability_%        816 non-null    int64  
 6   valence_%             816 non-null    int64  
 7   energy_%              816 non-null    int64  
 8   acousticness_%        816 non-null    int64  
 9   instrumentalness_%    816 non-null    int64  
 10  liveness_%            816 non-null    int64  
 11  speechiness_%         816 non-null    int64  
dtypes: float64(1), int64(11)
memory usage: 82.9 KB


In [9]:
#Selección de variables predictoras
x_base = df1[[
    'released_year','in_spotify_playlists','in_spotify_charts','bpm',
    'danceability_%', 'valence_%', 'energy_%',
    'acousticness_%', 'instrumentalness_%', 'liveness_%', 'speechiness_%'
]].values

y = df1['streams'].astype(float).values  # Asegúrate que y sea numérico
n = len(y)

In [10]:
# Escalamiento antes de aplicar potencias
scaler = StandardScaler()
x_scaled = scaler.fit_transform(x_base)

# Construcción de términos polinómicos manuales
x1 = x_scaled
x2 = x1**2
x3 = x1**3
unos = np.ones([n, 1])
X = np.hstack([unos, x1, x2, x3])

#Separo nuestros datos un 60%
x_train, x_test, y_train, y_test = train_test_split(X, y, test_size=0.60, random_state=137)

#estandarizamos
scaler=StandardScaler().fit(x_train)

x_train_scaled=scaler.transform(x_train)
x_test_scaled=scaler.transform(x_test)

### Realizamos el modelo sin Penalización

In [11]:
#Hacemos la regresión lineal
x_train_e= sm.add_constant(x_train_scaled)
resultado= sm.OLS(y_train,x_train_e).fit()
resultado.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                      y   R-squared:                       0.786
Model:                            OLS   Adj. R-squared:                  0.761
Method:                 Least Squares   F-statistic:                     32.43
Date:                Sun, 28 Sep 2025   Prob (F-statistic):           8.61e-79
Time:                        11:24:37   Log-Likelihood:                -6772.7
No. Observations:                 326   AIC:                         1.361e+04
Df Residuals:                     292   BIC:                         1.374e+04
Df Model:                          33                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const       4.795e+08   1.49e+07     32.152      0.000     4.5e+08    5.09e+08
x1          3.994e-07    3.9e-07      1.024      0.307   -3.68e-07    1.17e-06
x2         -2.769e+07   8.27e+07     -0.335      0.738    -1.9e+08    1.35e+08
x3          8.581e+08   5.25e+07     16.358      0.000    7.55e+08    9.61e+08
x4          7.257e+07   3.25e+07      2.231      0.026    8.55e+06    1.37e+08
x5          3.149e+06   2.76e+07      0.114      0.909   -5.12e+07    5.75e+07
x6          2.244e+07   3.07e+07      0.732      0.465   -3.79e+07    8.28e+07
x7          1.292e+07   3.46e+07      0.373      0.709   -5.53e+07    8.11e+07
x8         -1.684e+07   3.05e+07     -0.552      0.582   -7.69e+07    4.32e+07
x9           4.49e+07   3.06e+07      1.469      0.143   -1.53e+07    1.05e+08
x10        -2.036e+05   9.57e+07     -0.002      0.998   -1.89e+08    1.88e+08
x11          2.37e+07   2.84e+07      0.835      0.405   -3.22e+07    7.96e+07
x12        -4.694e+07    3.7e+07     -1.268      0.206    -1.2e+08    2.59e+07
x13        -4.168e+08   1.83e+08     -2.278      0.023   -7.77e+08   -5.68e+07
x14        -5.185e+08   1.26e+08     -4.104      0.000   -7.67e+08    -2.7e+08
x15        -1.662e+08   8.35e+07     -1.989      0.048   -3.31e+08   -1.78e+06
x16        -1.104e+07      3e+07     -0.368      0.713   -7.01e+07     4.8e+07
x17        -1.819e+07   2.23e+07     -0.817      0.414    -6.2e+07    2.56e+07
x18          2.17e+07   1.63e+07      1.331      0.184   -1.04e+07    5.38e+07
x19        -1.608e+07   2.06e+07     -0.782      0.435   -5.66e+07    2.44e+07
x20         2.417e+07   4.72e+07      0.513      0.609   -6.86e+07    1.17e+08
x21          9.94e+07   2.58e+08      0.385      0.701   -4.09e+08    6.08e+08
x22        -8.676e+07   8.85e+07     -0.981      0.328   -2.61e+08    8.74e+07
x23         4.843e+07   7.84e+07      0.618      0.537   -1.06e+08    2.03e+08
x24        -3.265e+08    1.2e+08     -2.712      0.007   -5.63e+08   -8.95e+07
x25         5.602e+07    9.9e+07      0.566      0.572   -1.39e+08    2.51e+08
x26         1.117e+08   6.77e+07      1.651      0.100   -2.15e+07    2.45e+08
x27         1.715e+07   3.89e+07      0.441      0.659   -5.94e+07    9.37e+07
x28         -3.58e+07   3.48e+07     -1.028      0.305   -1.04e+08    3.27e+07
x29         1.323e+07   3.35e+07      0.394      0.694   -5.28e+07    7.92e+07
x30         9.621e+06   3.12e+07      0.309      0.758   -5.17e+07     7.1e+07
x31        -3.709e+07   5.92e+07     -0.627      0.531   -1.54e+08    7.94e+07
x32         -9.59e+07   1.79e+08     -0.536      0.593   -4.48e+08    2.57e+08
x33         7.583e+07   7.62e+07      0.995      0.321   -7.42e+07    2.26e+08
x34        -1.354e+07    6.1e+07     -0.222      0.824   -1.34e+08    1.07e+08
==============================================================================
Om

In [12]:
#Sacamos lor R2 de x_train y x_test

#preparamos x_test
x_test_e=sm.add_constant(x_test_scaled)
y_pred=resultado.predict(x_test_e)
#calculamos r2 de xtrain
r2_train=resultado.rsquared
print("R2 de train: ",r2_train)

#calculamos r2 de xtest
r2_test=r2_score(y_test, y_pred)
print("R2 de test: ",r2_test)

R2 de train:  0.785640116926638
R2 de test:  0.713369034448787


### Con Penalización de L2, Ridge

In [13]:
#Con alfa = 0.5, con xtrain_scaled y xtest_scaled
ridge= Ridge(alpha=0.5)
ridge.fit(x_train_scaled, y_train)
#preparamos para sacar las r2 de ridge1 (nuestro modelo), sacando las y_pred, 
y_pred_train=ridge.predict(x_train_scaled)
y_pred_test=ridge.predict(x_test_scaled)

#sacamos r2
r2_ridge_train=r2_score(y_train,y_pred_train)
r2_ridge_test=r2_score(y_test,y_pred_test)
print("R2 de train (Ridge): ", r2_ridge_train)
print("R2 de test (Ridge): ", r2_ridge_test)

R2 de train (Ridge):  0.785116339739985
R2 de test (Ridge):  0.7121496860437078


### Con Penalización de L1, Lasso

In [14]:
#Con alfa = 0.5, con xtrain_scaled y xtest_scaled
lasso= Lasso(alpha=0.5)
lasso.fit(x_train_scaled, y_train)
#preparamos para sacar las r2 de ridge1 (nuestro modelo), sacando las y_pred, 
y_pred_trainl=lasso.predict(x_train_scaled)
y_pred_testl=lasso.predict(x_test_scaled)

#sacamos r2
r2_lasso_train=r2_score(y_train,y_pred_trainl)
r2_lasso_test=r2_score(y_test,y_pred_testl)
print("R2 de train (Lasso): ", r2_lasso_train)
print("R2 de test (Lasso) : ", r2_lasso_test)

R2 de train (Lasso):  0.7856400640789517
R2 de test (Lasso) :  0.7133603664916135


C:\Users\Lenovo\anaconda3\Lib\site-packages\sklearn\linear_model\_coordinate_descent.py:695: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations, check the scale of the features or consider increasing regularisation. Duality gap: 1.058e+19, tolerance: 9.877e+15
  model = cd_fast.enet_coordinate_descent(


### Con penalización ElasticNet

In [15]:
# Con alfa = 0.5, porque así se comporta como ridge y lasso
elastic = ElasticNet(alpha=0.5, l1_ratio=0.5)
elastic.fit(x_train_scaled, y_train)

# Predicciones
y_pred_train_elastic = elastic.predict(x_train_scaled)
y_pred_test_elastic = elastic.predict(x_test_scaled)

# Sacamos las R^2
r2_elastic_train = r2_score(y_train, y_pred_train_elastic)
r2_elastic_test = r2_score(y_test, y_pred_test_elastic)

print("R² de train (ElasticNet):", r2_elastic_train)
print("R² de test (ElasticNet):", r2_elastic_test)

R² de train (ElasticNet): 0.64661025120928
R² de test (ElasticNet): 0.594488794135108


### Comparamos las R^2

In [16]:
#R^2 de train
print("R2 de train (sin penalización): ",r2_train)
print("R2 de train (Ridge): ", r2_ridge_train)
print("R2 de train (Lasso): ", r2_lasso_train)
print("R² de train (ElasticNet):", r2_elastic_train)

R2 de train (sin penalización):  0.785640116926638
R2 de train (Ridge):  0.785116339739985
R2 de train (Lasso):  0.7856400640789517
R² de train (ElasticNet): 0.64661025120928


In [17]:
#R^2 de test
print("R2 de test: ",r2_test)
print("R2 de test (Ridge): ", r2_ridge_test)
print("R2 de test (Lasso) : ", r2_lasso_test)
print("R² de test (ElasticNet):", r2_elastic_test)

R2 de test:  0.713369034448787
R2 de test (Ridge):  0.7121496860437078
R2 de test (Lasso) :  0.7133603664916135
R² de test (ElasticNet): 0.594488794135108


Podemos observar que el R^2 de entrenamiento del modelo sin penalizar es el mejor, esto nos sugiere que es el modelo más preciso para identificar la relación entre las carecterísticas musicales y el éxito en stream. Sin embargo, al no penalizar nos podemos enfrentar al sobreajuste, por eso realizamos las penalizaciones de Lasso y Ridge, que nos dieron valores muy similares a nuestra R^2 sin penalizar, mientras que, en ElasticNet tuvimos una diferencia notable, esto puede ser porque este descartando variables útiles por exceso de regularización. Con estos resultados podemos decir que las características musicales que presentan las canciones si pueden explicar bien su éxito en Spotify.

### Análisis de significancia de factores para modelo sin penalización.

En el modelo podemos ver que presentamos p-values muy altos, con excepción de tres variables que mostraron si tener significancia, x3 y x4 que corresponde a una variable original, x13 y x14 que son valores cuadrados, y x24 que es un valor cúbico. Estos valores nos dicen que estan directamente relacionados con la popularidad o éxito de la canción, teniendo un efecto creciente no lineal sobre los streams, también en el caso de las variables x13, x14 y x24 podemos suponer que estas características de la canción entre más crezcan tienen más efecto en el éxito de una canción.
Por otra parte, observamos que el resto de las variables tienen p-values muy altos, lo que indica que sus efectos no son estadísticamente significativos en este modelo, dandonos a entender que el éxito musical de una canción esta ligado a un subconjunto específico de características, la popularidad musical no depende de una fórmula universal, sino de combinaciones específicas de características que atraen al público.  

### **Modelo 2: interacciones**

In [36]:
# Selección de variables predictoras (mismas que en Modelo 1)
x_base2 = df1[[
    'released_year','in_spotify_playlists','in_spotify_charts','bpm',
    'danceability_%', 'valence_%', 'energy_%',
    'acousticness_%', 'instrumentalness_%', 'liveness_%', 'speechiness_%'
]].values

y2 = df1['streams'].astype(float).values  # Asegúrate que y sea numérico
n2 = len(y2)

In [37]:
# Escalamiento antes de aplicar potencias
scaler2 = StandardScaler()
x_scaled2 = scaler2.fit_transform(x_base2)


In [38]:
# Construcción de términos polinómicos manuales (solo hasta grado 2)
x1_2 = x_scaled2
x2_2 = x1_2**2
unos2 = np.ones([n2, 1])
X2 = np.hstack([unos2, x1_2, x2_2])   

In [39]:
# Separo nuestros datos un 60%
x2_train, x2_test, y2_train, y2_test = train_test_split(X2, y2, test_size=0.60, random_state=137)

# estandarizamos
scaler2b = StandardScaler().fit(x2_train)
x2_train_scaled = scaler2b.transform(x2_train)
x2_test_scaled  = scaler2b.transform(x2_test)


In [40]:
# ==== Sin penalización ====
x2_train_e = sm.add_constant(x2_train_scaled)
resultado2 = sm.OLS(y2_train, x2_train_e).fit()
resultado2.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                      y   R-squared:                       0.776
Model:                            OLS   Adj. R-squared:                  0.760
Method:                 Least Squares   F-statistic:                     47.66
Date:                Sun, 28 Sep 2025   Prob (F-statistic):           8.42e-85
Time:                        23:13:19   Log-Likelihood:                -6780.0
No. Observations:                 326   AIC:                         1.361e+04
Df Residuals:                     303   BIC:                         1.369e+04
Df Model:                          22                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const       4.795e+08    1.5e+07     32.025      0.000     4.5e+08    5.09e+08
x1          9.606e-08   4.53e-08      2.119      0.035    6.84e-09    1.85e-07
x2          1.411e+08   5.02e+07      2.810      0.005    4.23e+07     2.4e+08
x3          8.754e+08   4.04e+07     21.659      0.000    7.96e+08    9.55e+08
x4          3.235e+07    2.6e+07      1.242      0.215   -1.89e+07    8.36e+07
x5          1.227e+07    1.7e+07      0.722      0.471   -2.12e+07    4.57e+07
x6         -1.365e+05   2.09e+07     -0.007      0.995   -4.12e+07    4.09e+07
x7          1.837e+07   1.88e+07      0.977      0.329   -1.86e+07    5.54e+07
x8         -8.765e+06   2.07e+07     -0.424      0.672   -4.95e+07    3.19e+07
x9          3.246e+07   2.34e+07      1.389      0.166   -1.35e+07    7.85e+07
x10         3.508e+07   4.54e+07      0.772      0.441   -5.44e+07    1.25e+08
x11         3.711e+06   2.44e+07      0.152      0.879   -4.42e+07    5.17e+07
x12        -6.082e+07   3.01e+07     -2.023      0.044    -1.2e+08   -1.67e+06
x13         5.373e+07   4.62e+07      1.164      0.245   -3.71e+07    1.45e+08
x14        -4.589e+08   3.46e+07    -13.256      0.000   -5.27e+08   -3.91e+08
x15        -3.189e+07   2.46e+07     -1.299      0.195   -8.02e+07    1.64e+07
x16         1.681e+06   1.84e+07      0.091      0.927   -3.46e+07     3.8e+07
x17        -6.493e+06   1.71e+07     -0.380      0.704   -4.01e+07    2.71e+07
x18         2.015e+07   1.58e+07      1.276      0.203   -1.09e+07    5.12e+07
x19        -2.365e+07   1.78e+07     -1.332      0.184   -5.86e+07    1.13e+07
x20        -1.017e+06   2.29e+07     -0.044      0.965    -4.6e+07     4.4e+07
x21        -2.949e+07   4.52e+07     -0.653      0.514   -1.18e+08    5.94e+07
x22         2.472e+06    2.4e+07      0.103      0.918   -4.48e+07    4.97e+07
x23         4.274e+07   2.86e+07      1.494      0.136   -1.36e+07     9.9e+07
==============================================================================
Omnibus:                      101.438   Durbin-Watson:                   2.048
Prob(Omnibus):                  0.000   Jarque-Bera (JB):              575.424
Skew:                           1.160   Prob(JB):                    1.12e-125
Kurtosis:                       9.081   Cond. No.                     3.00e+16
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The smallest eigenvalue is 1.18e-30. This might indicate that there are
strong multicollinearity problems or that the design matrix is singular.
"""

In [41]:


# R² de train y test
x2_test_e = sm.add_constant(x2_test_scaled)
y2_pred = resultado2.predict(x2_test_e)

r2_train2 = resultado2.rsquared
print("R2 de train (grado 2, sin penalización): ", r2_train2)

r2_test2 = r2_score(y2_test, y2_pred)
print("R2 de test  (grado 2, sin penalización): ", r2_test2)


R2 de train (grado 2, sin penalización):  0.7758013395273556
R2 de test  (grado 2, sin penalización):  0.7058031182841495


In [42]:
# ==== Ridge ====
ridge2 = Ridge(alpha=0.5)
ridge2.fit(x2_train_scaled, y2_train)
print("R2 de train (grado 2, Ridge):", r2_score(y2_train, ridge2.predict(x2_train_scaled)))
print("R2 de test  (grado 2, Ridge):", r2_score(y2_test,  ridge2.predict(x2_test_scaled)))


R2 de train (grado 2, Ridge): 0.7757038193351681
R2 de test  (grado 2, Ridge): 0.7075467587215022


In [43]:
# ==== Lasso ====
lasso2 = Lasso(alpha=0.5, max_iter=10000)
lasso2.fit(x2_train_scaled, y2_train)
print("R2 de train (grado 2, Lasso):", r2_score(y2_train, lasso2.predict(x2_train_scaled)))
print("R2 de test  (grado 2, Lasso):", r2_score(y2_test,  lasso2.predict(x2_test_scaled)))


R2 de train (grado 2, Lasso): 0.7758013395273555
R2 de test  (grado 2, Lasso): 0.7058031212336524


In [44]:
# ==== ElasticNet ====
elastic2 = ElasticNet(alpha=0.5, l1_ratio=0.5, max_iter=10000)
elastic2.fit(x2_train_scaled, y2_train)
print("R2 de train (grado 2, ElasticNet):", r2_score(y2_train, elastic2.predict(x2_train_scaled)))
print("R2 de test  (grado 2, ElasticNet):", r2_score(y2_test,  elastic2.predict(x2_test_scaled)))

R2 de train (grado 2, ElasticNet): 0.6287050130579301
R2 de test  (grado 2, ElasticNet): 0.5863604769607365


In [45]:
# COMPARACIÓN DE R² (Modelo 2)
print("\n===== R² de TRAIN (Modelo 2: grado 2) =====")
print("Sin penalización:", r2_train2)
print("Ridge:", r2_score(y2_train, ridge2.predict(x2_train_scaled)))
print("Lasso:", r2_score(y2_train, lasso2.predict(x2_train_scaled)))
print("ElasticNet:", r2_score(y2_train, elastic2.predict(x2_train_scaled)))

print("\n===== R² de TEST (Modelo 2: grado 2) =====")
print("Sin penalización:", r2_test2)
print("Ridge:", r2_score(y2_test, ridge2.predict(x2_test_scaled)))
print("Lasso:", r2_score(y2_test, lasso2.predict(x2_test_scaled)))
print("ElasticNet:", r2_score(y2_test, elastic2.predict(x2_test_scaled)))


===== R² de TRAIN (Modelo 2: grado 2) =====
Sin penalización: 0.7758013395273556
Ridge: 0.7757038193351681
Lasso: 0.7758013395273555
ElasticNet: 0.6287050130579301

===== R² de TEST (Modelo 2: grado 2) =====
Sin penalización: 0.7058031182841495
Ridge: 0.7075467587215022
Lasso: 0.7058031212336524
ElasticNet: 0.5863604769607365


#### **Analisis**

El Modelo 2 (grado 2) mostró un buen desempeño predictivo con un R² cercano a 0.77 en entrenamiento y 0.70 en prueba, resultados muy similares al modelo cúbico del primer caso. Los coeficientes obtenidos indican que solo algunos predictores resultan estadísticamente significativos (como x1, x2, x3, x12 y x14), mientras que la mayoría no aporta evidencia suficiente, lo que refleja que gran parte de los términos cuadráticos agregan ruido más que información. Además, los diagnósticos revelan problemas de multicolinealidad severa (cond. no. muy elevado) y colas pesadas en los residuos, aunque no se observó autocorrelación. Las versiones con Ridge y Lasso estabilizan el modelo sin modificar mucho el poder explicativo, mientras que ElasticNet con los parámetros probados pierde desempeño. En conclusión, este modelo confirma que no es necesario llegar a grado tres para capturar la relación entre las variables y los streams, pues con grado dos ya se alcanza un ajuste muy parecido, aunque con menor interpretabilidad debido a la presencia de muchos términos poco relevantes.